### DISCUSSION TITLE
Gosund SP1 flashed with Tasmota 9.5.0, Energy today values ​​implausible

### DISCUSSION THREAD

**Discussion by haselchen** [arendst/Tasmota#15749] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Gosund SP1
- [x] Tasmota binary firmware version number used: 9.5.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Tuya Convert
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
`18:42:09.064 MQT: stat/tasmota/RESULT = {"NAME":"Gosund SP1 v23","GPIO":[0,321,0,32,2720,2656,0,0,2624,320,224,0,0,0],"FLAG":0,"BASE":55}
18:42:09.278 MQT: stat/tasmota/RESULT = {"Module":{"0":"Gosund SP1 v23"}}
18:42:09.533 MQT: stat/tasmota/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"321":"Led_i2"},"GPIO2":{"0":"None"},"GPIO3":{"32":"Button1"},"GPIO4":{"2720":"BL0937 CF"},"GPIO5":{"2656":"HLWBL CF1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"2624":"HLWBL SEL_i"},"GPIO13":{"320":"Led_i1"},"GPIO14":{"224":"Relay1"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
18:42:16.985 MQT: tele/tasmota/STATE = {"Time":"2022-06-04T18:42:16","Uptime":"0T01:40:56","UptimeSec":6056,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":24,"MqttCount":2,"POWER":"ON","Wifi":{"AP":1,"SSId":"XXXX","BSSId":"XXXX","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":2,"Downtime":"0T00:00:06"}}
18:42:17.001 MQT: tele/tasmota/SENSOR = {"Time":"2022-06-04T18:42:16","ENERGY":{"TotalStartTime":"2022-04-23T15:13:47","Total":4.954,"Yesterday":0.000,"Today":1.015,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":230,"Current":0.000}}`

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

`18:44:12.197 MQT: stat/tasmota/STATUS = {"Status":{"Module":0,"DeviceName":"StromverbrauchKühlschrankTerrasse","FriendlyName":["StromverbrauchKühlschrankTerrasse"],"Topic":"tasmota","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:44:12.233 MQT: stat/tasmota/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://thehackbox.org/tasmota/release/tasmota.bin","RestartReason":"Software/System restart","Uptime":"0T01:42:52","StartupUTC":"2022-06-04T15:01:20","Sleep":50,"CfgHolder":4617,"BootCount":156,"BCResetTime":"2021-06-09T17:02:13","SaveCount":524,"SaveAddress":"F8000"}}
18:44:12.267 MQT: stat/tasmota/STATUS2 = {"StatusFWR":{"Version":"9.5.0(tasmota)","BuildDateTime":"2021.06.17 08:28:30","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285","CR":"518/699"}}
18:44:12.286 MQT: stat/tasmota/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["WLAN-MaMoMu",""],"TelePeriod":10,"Resolution":"558180C0","SetOption":["00008009","2805C8000100068000005A0A000000000000","00000080","00006000","00000000"]}}
18:44:12.317 MQT: stat/tasmota/STATUS4 = {"StatusMEM":{"ProgramSize":608,"Free":392,"Heap":25,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144051","FlashFrequency":40,"FlashMode":3,"Features":["00000407","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
18:44:12.349 MQT: stat/tasmota/STATUS5 = {"StatusNET":{"Hostname":"SteckdoseKühlschrankTerrasse","IPAddress":"XXXX","Gateway":"XXXX","Subnetmask":"255.255.255.0","DNSServer":"XXXX","Mac":"XXXX","Webserver":2,"WifiConfig":2,"WifiPower":17.0}}
18:44:12.372 MQT: stat/tasmota/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.2.49","MqttPort":8883,"MqttClientMask":"StromverbrauchKühlschrankTerrasse","MqttClient":"StromverbrauchKühlschrankTerrasse","MqttUser":"XXXX","MqttCount":2,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:44:12.401 MQT: stat/tasmota/STATUS7 = {"StatusTIM":{"UTC":"2022-06-04T16:44:12","Local":"2022-06-04T18:44:12","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":99,"Sunrise":"05:49","Sunset":"21:47"}}
18:44:12.417 MQT: stat/tasmota/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
18:44:12.434 MQT: stat/tasmota/STATUS10 = {"StatusSNS":{"Time":"2022-06-04T18:44:12","ENERGY":{"TotalStartTime":"2022-04-23T15:13:47","Total":4.954,"Yesterday":0.000,"Today":1.015,"Power":3,"ApparentPower":3,"ReactivePower":0,"Factor":1.00,"Voltage":230,"Current":0.013}}}
18:44:12.458 MQT: stat/tasmota/STATUS11 = {"StatusSTS":{"Time":"2022-06-04T18:44:12","Uptime":"0T01:42:52","UptimeSec":6172,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":31,"MqttCount":2,"POWER":"ON","Wifi":{"AP":1,"SSId":"WLAN-MaMoMu","BSSId":"3C:37:12:0E:C3:62","Channel":11,"Mode":"11n","RSSI":84,"Signal":-58,"LinkCount":2,"Downtime":"0T00:00:06"}}}
18:44:16.973 MQT: tele/tasmota/STATE = {"Time":"2022-06-04T18:44:16","Uptime":"0T01:42:56","UptimeSec":6176,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"POWER":"ON","Wifi":{"AP":1,"SSId":"XXXX","BSSId":"XXXX","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":2,"Downtime":"0T00:00:06"}}
18:44:17.000 MQT: tele/tasmota/SENSOR = {"Time":"2022-06-04T18:44:16","ENERGY":{"TotalStartTime":"2022-04-23T15:13:47","Total":4.954,"Yesterday":0.000,"Today":1.015,"Period":0,"Power":3,"ApparentPower":3,"ReactivePower":0,"Factor":1.00,"Voltage":230,"Current":0.011}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

`Hi Everbody,
I measure the current of a refrigerator. I flashed a Gosund Sp1 with Tasmota 9.5.0
The socket has been calibrated with an ammeter.
However, the energy values today do not match the consumption in any way.
In the Tasmota commands, Kh is written.
But the GUI says kWh. What is right?
As an example I have the refrigerator, which consumes about 50 W.
The energy today value literally shoots up every second.
It ran for almost 3 hours with breaks and is said to have consumed 1 kWh.
`

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

![grafik](https://user-images.githubusercontent.com/42101470/172017147-a92a03b5-a157-4ea8-adce-d5d5df8e7fc2.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment, marked as answer]
What is implausible ?
I found multiple sources on the Internet say that the overall yearly power consumptiion of a fridge is between 200 to 400kWh per year.
So 1 kWh on a day looks reasonnable.

**Reply by ascillato2** [reply]
@haselchen

50 W/h for a fridge is ok if it is a small refrigerator or have inverter technology. Is it?

For example, mine is a common one. Not so new, not so old and non inverter. It consumes 150W to 250W. The measurements shows consumption cycles of about one hour in 150W and half an hour in 250W. It varies but it is more less that. The Power factor is .7 (and that is normal for a fridge compressor). So, 1Kw/day is a good value for your fridge. Mine is around 2 to 3 Kw/day.

Remember that the power (P) in W is different to the apparent Power (S) due to the power factor. And also, take into account that it may have spikes of consumption too.

**Reply by haselchen** [reply]
@barbudor 

I think we misunderstand each other.
The fridge doesn't run at 50 watts all the time. Sometimes it turns off by itself.
It can't be that I used 1kWh after 6.5 hours (with no less than 50 watts)
This is either a naming error in the GUI or a software bug.

It's a new economical refrigerator.
127 kWh per year.

**Reply by barbudor** [reply]
Yes kWh is what is intended
Try with a fixed consumption device like a halogen light
If the device has a very changing power consumption (like within the same second) , it may be possible that sampling of measure don't fit well.

**Comment by ascillato2** [comment]
Please, update to latest Tasmota version. Your version is very old and a huge amount of bug fixing has been done since your version. Please, try again but with latest Tasmota. Thanks.

**Reply by haselchen** [reply]
I've tried all versions from 11 onwards.
The question is whether kWh is the correct term for Energy today. The calculation is wrong.

**Reply by sfromis** [reply]
KWh is indeed the correct unit for Tasmota energy monitoring. I've got many energy monitoring devices, which all shows entirely plausible values for the devices they are monitoring, and the total KWh I get aligns very well with what the supplier charges me for. I've seen no issues with the numbers since I started with that monitoring on Tasmota 6.5, and updating through releases since then. My fridge currently use around 200-250 Wh a day.

With what you see, I'd do the calibration again, making sure to do the steps exactly right, and use a purely resistive load. One "trap" is that the current is to be input in milliampere, not the usual ampere values.

**Reply by haselchen** [reply]
I have performed the calibration x times.
As I said, the wattage is correct.
But not the calculation.
As an example, I had a TV with 155 watts on the socket for 5 hours. The GUI shows me 16!!! kWh.
According to the electricity meter, we only used 10 in the whole day.

**Reply by arendst** [reply]
I guess another template problem.

Try to change template gpio4 from ``BL0937 CF`` to ``HLW8012 CF``. Then re-calibrate and retry.

**Reply by haselchen** [reply]
If that helps, that would be great.
I have 15 of these sockets.
Would the template then have to be adjusted from your side?
I work with IOBroker and would then post the solution there in the forum.

**Reply by arendst** [reply]
Manufacterers build all kinds of energy hardware into devices. I see your Gosund SP1 is using a different template then the original default template for a reason unknown to me.

There are currently over 2300 different templates available in addition to the 75 fixed templates. Modern templates are user configrable and downloadable so no need for me to do anything. You just need to find the right template for your hardware.

You might even want to try the default template by executing command ``module 55`` and see what happens.

**Reply by haselchen** [reply]
I had already tried all of them.
The tip with the GPIO4 change is probably worth its weight in gold.
I've read about HLW8012 hardware related to Gosund sockets many times, but thought nothing of it.
I will monitor and report the values.
If I see it correctly so far, it could fit.
100W lamp consumes 0.12 kWh in 1 hour. From a mathematical point of view, that would fit.

**Reply by haselchen** [reply]
The workaround seems to have worked. However, you then have to calibrate again very precisely. And very important, the device must be plugged into the socket. Entering the set commands in the console without a device being plugged in brings measurement errors.
Right or wrong, I've set SetOption72 to 1.

Please leave the issue open, I would like to do a test over a longer period of time.

**Reply by sfromis** [reply]
If the issue was that you did not have a configuration to match the specific energy monitoring chip in your hardware, the change to fix that was no workaround, it was just correcting an error. 

Yes, calibration is needed when you change energy monitoring chip config, that is entirely unsurprising. And you cannot repeat the calibration commands from one device to the next, you **have to** to do a new calibration for each individual device, as components may have varying values. Thus it would make no sense to enter the commands without being in the process of running a new calibration on the particular device.

**Reply by haselchen** [reply]
Practically I did it like this:
Multiple socket, in it the Gosund socket and a current measuring socket with lamp.
I had read the values ​​from the measuring socket and entered them into the Gosund's console.
After I then put the lamp in the Gosund, the GUI showed 140W instead of 100W. I found that amazing.
So again entered the values ​​when the lamp was connected.
And then the result was satisfactory. In each test, the Gosund read 98-100W. And Energy Today also seems to be calculated correctly now.

**Comment by haselchen** [comment]
@barbudor 
@arendst 
@ascillato2 
@sfromis 

This Issue can be closed.
The tip with GPIO4 was the solution. I can reasonably understand the values ​​I'm getting now.
Thanks for your advice and time.
